### ISSUE TITLE
🐛 [BUG] - The system uses 60hz instead of using the highest refresh rate the monitor provides (mine is 100hz)

### ISSUE THREAD

**Issue by Hirakamu** [BlissRoms-x86/support#106]
### Description

Yes, i don't know if something necessary like this has to be reported to GitHub, but this is very annoying for me and I can't find a solution to my personal problem

### Reproduction steps

```bash
1. Install a new fresh copy of BlissOS
2. Activate developer mode
3. Turn on 'Show refresh rate' options
```

### Screenshots

```bash
I don't know how to take a screenshot by default in Bliss, the important thing is that a red number appears on the screen and says '60' (I know this is trivial, but I just followed the instructions)
```

### Logs

```bash
why?
```

### Category

Display

### OS Version

16.x

**Comment by Hirakamu** [comment]
I just want to enjoy BlissOS with a high refresh rate

**Comment by Xtr126** [comment]
The issue is that the default mode your monitor advertises to the OS (Bliss in this case) is the one with lower refresh rate.
Then hwcomposer picks it first and stuck with it.
It’s possible to tell drm_minigbm_celadon hwcomposer to use a specific mode by using kernel parameters, but this requires figuring out mode id and connector id which you can do only if you are familiar with linux terminal and kernel (graphics) drm subsystem in sysfs/dev. 

For celadon hwc you can “select monitor” with these experimental flags, MODE_ID and CONNECTOR_ID https://github.com/BlissRoms-x86/device_generic_common/blob/6017b7d8071e148623efca41531fc1c5ff2a6092/init.sh#L498
This is how I got it to work long ago 
Run this command `cd /sys/class/drm; for i in card*; do echo $i; cat $i/modes; done `
And get an output:
```sh
console:/sys/class/drm # cd /sys/class/drm; for i in card*; do echo $i; cat $i/modes; done 
card0
cat: card0/modes: No such file or directory
card0-Virtual-1
5120x2160
4096x2160
3840x2160
3840x2160
3840x2160
1920x1440
2560x1080
1856x1392
1792x1344
2048x1152
1920x1200
1920x1080
1920x1080
1600x1200
1680x1050
1400x1050
1280x1024
1440x900
1280x960
1360x768
1280x768
1024x768
800x600
640x480
640x480
console:/sys/class/drm # 
```
I used QEMU for this demo but it also works on physical devices.
If you want the second resolution, 4096x2160
then MODE_ID=2
Then obtain CONNECTOR_ID from `logcat | grep hwc` (as soon as android boots before logs are erased)
For example from here it would be CONNECTOR_ID=34
```sh
130|console:/ # logcat | grep hwc                                              
11-14 06:45:28.412   659   659 D hwc-drm-device: The property 'vendor.hwcomposer.preferred.mode.limit' value is 1
11-14 06:45:28.412   659   659 D hwc-drm-device: The property 'vendor.hwcomposer.planes.enabling' value is 0, only support primary plane
11-14 06:45:28.412   659   659 D hwc-drm-device: The property 'vendor.hwcomposer.planes.num' value is 99
11-14 06:45:28.413   659   659 I hwc-buffer-info-getter: Using CrOS Gralloc gralloc module: Chrome OS
11-14 06:45:28.413   659   659 D hwc-drm-connector: The property 'vendor.hwcomposer.connector.id' value is -1
11-14 06:45:28.413   659   659 D hwc-drm-connector: The property 'vendor.hwcomposer.mode.id' value is -1
11-14 06:45:28.413   659   659 D hwc-drm-connector: CONNECTOR:34 select one mode, id = 1, name = 1920x1080@59.993221, refresh = 59.993221
11-14 06:45:28.413   659   659 D hwc-drm-connector: CONNECTOR:34 preferred mode found, set preferred mode id = 1, name = 1920x1080@59.993221, refresh = 59.993221
11-14 06:45:28.413   659   659 D hwc-drm-connector: The property 'vendor.hwcomposer.connector.multi_refresh_rate' value is -1
11-14 06:45:28.413   659   659 I hwc-resource-manager: Attaching connector Virtual-1
11-14 06:45:28.413   659   659 I hwc-drm-two: Attaching pipeline 'Virtual-1' to the display #0 (Primary)
11-14 06:45:28.413   659   659 D hwc-drm-connector: The property 'vendor.hwcomposer.connector.id' value is -1
11-14 06:45:28.413   659   659 D hwc-drm-connector: The property 'vendor.hwcomposer.mode.id' value is -1
11-14 06:45:28.413   659   659 D hwc-drm-connector: CONNECTOR:34 select one mode, id = 1, name = 1920x1080@59.993221, refresh = 59.993221
11-14 06:45:28.413   659   659 D hwc-drm-connector: CONNECTOR:34 preferred mode found, set preferred mode id = 1, name = 1920x1080@59.993221, refresh = 59.993221
11-14 06:45:28.413   659   659 D hwc-drm-connector: The property 'vendor.hwcomposer.connector.multi_refresh_rate' value is -1
11-14 06:45:28.414   659   659 I hwc-backend: Backend 'generic' for 'Virtual-1' and driver 'virtio_gpu' was successfully set
```
On my laptop MULTI_REFRESH_RATE=1 in kernel cmdline was also sufficient to let me switch between different refresh rates from Android settings > display.

Switching to gbm_gralloc might work as an alternative if you cannot find modes/connector, however it may cause other issues.
```
GRALLOC=gbm HWC=drm_gbm_cros drm_mode=1920x1080@100 video=1920x1080@100
```
TL;DR: You have find out the mode id and connector id for 100hz and supply them in kernel cmdline with MODE_ID= and CONNECTOR_ID= and use HWC=drm_minigbm_celadon GRALLOC=minigbm.

Another solution is to use edid_override feature of Linux kernel, it must be triggered before Android starts to use the GPU.

**Comment by Moahmme008** [comment]
how to modify the resolution and refresh rate of bliss??

**Comment by flexbrionine** [comment]
> how to modify the resolution and refresh rate of bliss??

Firstly you need to go into GNU Grub 2 boot settings by pressing "e" ob bliss boot menu when powering on (black screen with boot parameters will appear). To modify resolution you simply need to **type** `GRALLOC=gbm HWC=drm_gbm_cros drm_mode=1920x1080@100 video=1920x1080@100` where "1920x1080" is the resolution you want to set and "100" is your refresh rate. **IMPORTANT!** make sure you typed this option as last argument in line: `linux $kd/... $src $@` so this line looks like that (example for 2k and 144Hz):
```bash
                linux $kd/kernel_a stack_depot_disable=on cgroup_disable=pressure r\
oot=/dev/ram0 androidboot.slot_suffix=_a $src $@ GRALLOC=gbm HWC=drm_gbm_cros drm_m\
ode=2560x1440@144 video=2560x1440@144
```
(you don't need to type "\" symbol)
Edit: to save and exit pres Ctrl+X
works for _Bliss OS 16.9.x x86_64-v2_ 👍

**Comment by Sheepmister** [comment]
> > how to modify the resolution and refresh rate of bliss??
> 
> Firstly you need to go into GNU Grub 2 boot settings by pressing "e" ob bliss boot menu when powering on (black screen with boot parameters will appear). To modify resolution you simply need to **type** `GRALLOC=gbm HWC=drm_gbm_cros drm_mode=1920x1080@100 video=1920x1080@100` where "1920x1080" is the resolution you want to set and "100" is your refresh rate. **IMPORTANT!** make sure you typed this option as last argument in line: `linux $kd/... $src $@` so this line looks like that (example for 2k and 144Hz):
> 
>                 linux $kd/kernel_a stack_depot_disable=on cgroup_disable=pressure r\
> oot=/dev/ram0 androidboot.slot_suffix=_a $src $@ GRALLOC=gbm HWC=drm_gbm_cros drm_m\
> ode=2560x1440@144 video=2560x1440@144
> (you don't need to type "" symbol) Edit: to save and exit pres Ctrl+X works for _Bliss OS 16.9.x x86_64-v2_ 👍

i tried this code and got stuck in boot                 linux $kd/kernel_a stack_depot_disable=on cgroup_disable=pressure r\
oot=/dev/ram0 androidboot.slot_suffix=_a $src $@ GRALLOC=gbm HWC=drm_gbm_cros drm_m\
ode=2560x1440@144 video=2560x1440@144

**Comment by Sheepmister** [comment]
any fix my monitor is 75refreshrate
